### PR TITLE
Enable multiple-level fold section of R document like RStudio

### PR DESCRIPTION
**Pull request by Yunuuuu**
Here is some example code to explain this work:
```
# section one ----
a <- 1:10

## first sub-section ----
b <- letters

### first sub-sub-section ----
d <- 1:2

## second sub-section ----
e <- LETTERS

### sub-sub-section ----
f <- 1:3

#### sub-sub-sub-section ----
g <- 1:20

# complex nested fold section can work well -----
a <- function(x) {b <- function() {
        # collapse following section will not go beyond this function `b`
        # section for function b ------
        x <- 1:2
        x
    }

    # section one ----
    a <- 1:10

    ## first sub-section ----
    b <- letters

    x <- function() {
        # section for function x -----
        c(a, b, d)
    }
}
```
image
This is what the outline look like:
image

### CODE REVIEW

File: R/folding-and-symbols-helper.r [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+            type = "block",
+            start_line = start_line,
+            end_line = end_line - 1L
+        )
+    }, block_lines_list, NULL)
+    c(section_ranges, block_ranges)
+}
+
+# for a r document, we split the document into two element, one for document
+# content out of blocks, another for document content in blocks.
```
Comment (Yunuuuu) [inline R/folding-and-symbols-helper.r:67]: blocks ranges are codes between pairs like ("[" and "]"), ("(" and ")"), and ("{" and "}").

File: R/folding-and-symbols-helper.r [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+# ** For document content in blocks, we only need reuse this function to treat
+# each blocks as a new r document.
+get_r_document_sections_rec <- function(start_line, end_line, doc_content, block_lines_list) {
+    if (start_line > end_line) {
+        return(NULL)
+    }
+    line_seq <- start_line:end_line
+    # only keep blocks between start_line and end_line
+    block_lines_list <- lapply(
+        block_lines_list, "[",
```
Comment (Yunuuuu) [inline R/folding-and-symbols-helper.r:79]: block_lines_list is a `list` with two element, the first element is the block start lines and the second is the block end lines

File: R/folding-and-symbols-helper.r [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+            block_lines_list[[1L]] == start_line &
+                block_lines_list[[2L]] - 1L == end_line
+        )
+    )
+
+    lines_out_blocks <- extract_lines_out_blocks(
+        line_seq, block_lines_list
+    )
+    sections_in_blocks <- NULL
+    if (length(block_lines_list) && length(block_lines_list[[1L]])) {
```
Comment (Yunuuuu) [inline R/folding-and-symbols-helper.r:91]: if there exists some blocks, we firstly extract the outermost blocks (`highest_level_block_lines`) and treat the content in  these blocks as a new R document, so we can re-use this function to parse them.

**Comment by renkun-ken** [comment]
Thanks for re-submitting the PR. I'll take a look later today.

**Comment by renkun-ken** [comment]
I re-arranged some functions so that `section.R` only contains shared functions, and renamed some functions so that they are more consistent with the rest.

**Comment by randy3k** [comment]
Thank you for working on it!

**Comment by renkun-ken** [comment]
@Yunuuuu Some of the test cases failed. Would you like to update the test cases in the following files according to the errors reported at <https://github.com/REditorSupport/languageserver/runs/7883213618?check_suite_focus=true>:

* https://github.com/REditorSupport/languageserver/blob/master/tests/testthat/test-symbol.R
* https://github.com/REditorSupport/languageserver/blob/master/tests/testthat/test-folding.R

File: R/folding.R [hunk 4]
```diff
@@ -105,7 +105,7 @@ get_document_blocks <- function(xdoc) {
         list(
             type = "block",
             start_line = start_line,
-            end_line = end_line
+            end_line = end_line - 1L
```
Comment (Yunuuuu) [inline R/folding.R:108]: Exclude lines where close brackets stand

File: tests/testthat/test-folding.R [hunk 5]
```diff
@@ -117,9 +121,9 @@ test_that("Folding range works in Rmarkdown", {
...
     expect_equal(result[[3]]$endLine, 12)
     expect_equal(result[[4]]$startLine, 4)
-    expect_equal(result[[4]]$endLine, 5)
+    expect_equal(result[[4]]$endLine, 5) # mark
     expect_equal(result[[5]]$startLine, 7)
     expect_equal(result[[5]]$endLine, 8)
     expect_equal(result[[6]]$startLine, 9)
-    expect_equal(result[[6]]$endLine, 10)
+    expect_equal(result[[6]]$endLine, 10) # mark
 })
```
Comment (Yunuuuu) [inline tests/testthat/test-folding.R:129]: By excluding lines with close brackets, this can be fixed

**Comment by Yunuuuu** [comment]
Thanks for the reviewing @renkun-ken, I have run `R CMD check` in my computer which signal no errors

File: R/folding.R (hunk 4 above)
Comment (Yunuuuu) [reply R/folding.R:108]: This is what r documents do

**Comment by renkun-ken** [comment]
In addition to updating existing test cases, are you interested in adding some new test cases for the newly introduced features in this PR?

**Comment by Yunuuuu** [comment]
Sorry for the late reply @renkun-ken, I'll improve the test tomorow, it's a little busy these days.

**Comment by Yunuuuu** [comment]
For symbols, the "Document section symbol works" test in test-symbol.R file has provide nested section symbol test

**Comment by Yunuuuu** [comment]
Sorry for the delayed test commits, it's busy for the new school semester,

**Review by renkun-ken — APPROVED** [review]
LGTM Thanks for working on this!

**Pull request merged**